fabtests/component/dmabuf-rdma/util.c

```diff
@@ -126,14 +126,32 @@ void sync_tcp(int sockfd)
 	(void)! read(sockfd, &dummy2, sizeof dummy2);
 }
 
+static int readFully(int fd, void *obuf, int len)
+{
+	int bytesLeft = len;
+	char *buf = (char *) obuf;
+	int bytesRead = 0;
+
+	while (bytesLeft > 0 && (bytesRead = read(fd, (void *) buf,
+		bytesLeft)) > 0) {
+		bytesLeft -= bytesRead;
+		buf += bytesRead;
+	}
+
+	if (bytesRead <= 0)
+		return bytesRead;
+
+	return len;
+}
+
 int exchange_info(int sockfd, size_t size, void *me, void *peer)
 {
 	if (write(sockfd, me, size) != size) {
 		fprintf(stderr, "Failed to send local info\n");
 		return -1;
 	}
 
-	if (read(sockfd, peer, size) != size) {
+	if (readFully(sockfd, peer, size) != size) {
```

Comment:
If this is a nonblocking socket exchanging data of a fixed size, use recvmsg(..., MSG_WAITALL).  Same applies to the write() call, which can write fewer bytes than size.

Reply:
It is a blocking socket, but sometimes the read can still return before reading all the data.

Reply:
Sorry, if this is nonblocking, write is fine.  You can also just use recv() instead of recvmsg().

Reply:
Agree that a similar call for write is also preferred, even though the error hasn't been encountered so far.

Reply:
And in my comments above, I meant *blocking* not nonblocking... sheesh, I need sleep.  It's the MSG_WAITALL flag that's needed for the recv() call.

Reply:
@Juee14Desai Please try the recvmsg() approach first to see if it fixes the issue.

Reply:
I made the recv() changes. CI is passing in my private build with this change.